app/src/full/kotlin/tunnel/DnsProxy.kt

```diff
@@ -120,6 +129,34 @@ internal class DnsProxy(
         }
     }
 
+    private fun forwardTls(destination: InetSocketAddress, rawData : ByteArray, originEnvelope: Packet?) {
+
+        var s : SSLSocket? = null
+        Result.of {
+            val s = SSLSocketFactory.getDefault().createSocket() as SSLSocket
+            s.connect(destination)
+            if (!HttpsURLConnection.getDefaultHostnameVerifier().verify(destination.hostName, s.session)) {
+                w("Hostname mismatch on DNS SSL connection")
+                throw SSLHandshakeException("Expected ${destination.hostName}, found ${s.session.peerPrincipal} ")
+            }
+
+            DataOutputStream(s.outputStream).use {
+                //send TCP request
+                it.writeShort(rawData.size)
+                it.write(rawData)
+                it.flush()
+            }
+
+            if (originEnvelope != null) forwarder.add(s, originEnvelope)
+            else Result.of { s.close() }
```

Comment:
instead of closing the socket if the envelope is `null` it would be better if the socket wouldn't even be created.

Reply:
I could make that change, but that would differ from how it's being done for the UDP request. Plus, currently it will send the request, just not add it to the forwarder if the envelope is null.

Reply:
Nevermind, I misunderstood. The nullcheck is only used for empty UDP packets, which doesn't occur here. I've removed the nullcheck.